src/core/fstream.cc

```diff
@@ -365,6 +368,20 @@ input_stream<char> make_file_input_stream(
     return make_file_input_stream(std::move(f), 0, std::move(options));
 }
 
+class closeable_file_data_source_impl : public file_data_source_impl {
+public:
+    closeable_file_data_source_impl(file f, uint64_t offset, uint64_t len, file_input_stream_options options):file_data_source_impl(std::move(f),  offset,  len,  std::move(options)) {
+
+    }
+    future<> close() override {
+        co_await file_data_source_impl::close();
+        co_return co_await _file.close();
+    }
+};
+
+data_source make_closeable_file_data_source(file f, uint64_t offset, uint64_t len, file_input_stream_options opt) {
```

Comment:
It's better to put `bool_class<something> close_file>` bit into `file_input_stream_options`